From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tao Xiaobai <1599170798@qq.com>
Date: Fri, 2 Oct 2026 16:51:49 +0800
Subject: [PATCH] Avoid comparing derived ItemLore lines

Compare and hash the canonical lore lines only instead of traversing the styled view derived from those same lines a second time.

Adapted from Paper commit f6862603ddc15b1d2a58eabe28695e23fea5446b.

diff --git a/net/minecraft/world/item/component/ItemLore.java b/net/minecraft/world/item/component/ItemLore.java
index 406985f7b38cf5f78ee01b427a756af0ffdbc94c..01a78282f1efe515223b44ede58cbacece254faa 100644
--- a/net/minecraft/world/item/component/ItemLore.java
+++ b/net/minecraft/world/item/component/ItemLore.java
@@ -40,6 +40,17 @@ public record ItemLore(List<Component> lines, List<Component> styledLines) imple
         return new ItemLore(Util.copyAndAdd(this.lines, component));
     }
 
+    // Paper - f6862603 - styledLines is derived from lines and should not be compared twice
+    @Override
+    public boolean equals(final Object object) {
+        return object == this || object instanceof ItemLore itemLore && this.lines.equals(itemLore.lines);
+    }
+
+    @Override
+    public int hashCode() {
+        return this.lines.hashCode();
+    }
+
     @Override
     public void addToTooltip(
         final Item.TooltipContext context, final Consumer<Component> consumer, final TooltipFlag flag, final DataComponentGetter components
